### PR TITLE
Default qutrit mixed execute

### PR DESCRIPTION
**Pull request by Gabriel-Bottrill**
**Context:**
This is the summation of the qutrit mixed device development, this device will be used to run noisy qutrit circuits. It uses the new device API.

**Description of the Change:**
Adds the execute method to DefaultQutritMixed, also sets device supports backprop derivative to True

**Benefits:**
Allows for noisy qutrit circuits to be executed.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: pennylane/devices/default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [inline pennylane/devices/default_qutrit_mixed.py:246]: I don't add a docstring as it should be the same as device_api's one.

File: pennylane/devices/default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [inline pennylane/devices/default_qutrit_mixed.py:279]: I added code for debugging as I have implemented it throughout the other PRs. If I should leave it out, please let me know.

File: pennylane/devices/default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [inline pennylane/devices/default_qutrit_mixed.py:77]: simulator_tracking and single_tape_support modifiers are added. 

I am not sure if I should add simulator tracking as the code that computes shots is not necessarily true as it doesn't take into account commuting GellMann measurements, instead it counts it naively. This is similar to how the code is run on the device, but may not be the best for an actual device as you may want to be more careful with the samples taken.

If anyone has an opinion on if I should rewrite this, remove it, or leave it as is I would appreciate the input.

File: setup.py
Comment (Gabriel-Bottrill) [inline setup.py:62]: I added the device to the setup.py file as it should be ready to use and it makes it consistent with other devices.

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [inline tests/devices/test_default_qutrit_mixed.py:28]: These three functions are used as helpers

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [inline tests/devices/test_default_qutrit_mixed.py:303]: Like with `test_default_qubit.py` I reuse tests from sampling, simulate, and measurement to make sure functionality doesn't change through device.

**Review by Gabriel-Bottrill — COMMENTED** [review]
I made sure to read through all my docstrings for consistency, hopefully this will be an easier PR to review now.

Also, here are some comments explaining parts of the PR and asking questions I needed clarified. Thank you for reviewing!

**Comment by Gabriel-Bottrill** [comment]
Hi @albi3ro, @glassnotes, @mudit2812, and @trbromley. This is it! The execute PR, this PR implements execute and makes the `default.qutrit.mixed` device usable (maybe useless until channels are added). I have added some comments above, hopefully it looks good.

File: tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py
Comment (glassnotes) [inline tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py:179]: Since these are all related tests, they can be included together in a new class.

File: tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py
Comment (glassnotes) [inline tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py:252]:
```suggestion
            ["Creating QNode(func=<function test_execution_debugging>)"],
```
Not sure about this but there is a missing ) and > here

File: tests/devices/test_default_qutrit_mixed.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit_mixed.py:33]: Can group these all into one class that focuses on testing generic device properties

File: tests/devices/test_default_qutrit_mixed.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit_mixed.py:177]:
```suggestion
        """Tests execution and gradients of a basic circuit using autograd."""
```

File: tests/devices/test_default_qutrit_mixed.py
Comment (glassnotes) [reply tests/devices/test_default_qutrit_mixed.py:299]: This is a lot of duplication, if nothing else, does it make sense to just inherit from that class so that you don't have to duplicate all the static methods? (PL team: is that fine convention-wise? other suggestions here?)

File: tests/devices/test_default_qutrit_mixed.py
Comment (glassnotes) [inline tests/devices/test_default_qutrit_mixed.py:613]: What a nasty operation :smiling_imp:

**Review by glassnotes — CHANGES_REQUESTED** [review]
@Gabriel-Bottrill I can see the light! Thanks for all your work on this.

Some small comments within, two major questions I'm leaving top-level:
 - Is there anything that can be done to reduce duplication of the code/tests?
 - At least from what I can see here, it seems a set of broader integration tests is missing, i.e. tests that do

```python
dev = qml.device("default.qutrit.mixed", wires=3)

@qml.qnode(dev)
...
```

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:613]: I was trying to implement something similar to the qubit version here, but the controlled qubit stuff is a bit past the controlled qutrit stuff so I implemented this.

File: tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py
Comment (Gabriel-Bottrill) [reply tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py:252]: It's supposed to be like this, the actual message includes the pointer to the function `Creating QNode(func=<function test_execution_debugging.<locals>.circuit at 0x7c2ed9c63c40> ...` Along with some other stuff.

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:33]: Yes, that's a good idea, thank you!

**Comment by Gabriel-Bottrill** [comment]
@glassnotes I will look at the duplication again tomorrow

File: doc/releases/changelog-dev.md
Comment (trbromley) [reply doc/releases/changelog-dev.md:292]: Thanks, yes it would be great to have a single shiny changelog entry that shows off `default.qutrit.mixed`, with all the composing PRs listed together.

File: pennylane/devices/default_qutrit_mixed.py
Comment (trbromley) [inline pennylane/devices/default_qutrit_mixed.py:79]:
```suggestion
    """A PennyLane Python-based device for mixed-state qutrit simulation.
```

**Review by trbromley — COMMENTED** [review]
Thanks @Gabriel-Bottrill!

File: pennylane/devices/default_qutrit_mixed.py
Comment (mudit2812) [reply pennylane/devices/default_qutrit_mixed.py:77]: @albi3ro could you help out here? I'm still a bit unfamiliar with how the tracker works.

File: pennylane/devices/default_qutrit_mixed.py
Comment (mudit2812) [inline pennylane/devices/default_qutrit_mixed.py:201]:
```suggestion
        if execution_config is None or execution_config.gradient_method in {"backprop", "best"}:
            if circuit is None:
                return True
            return not circuit.shots and not any(
                isinstance(m.obs, qml.SparseHamiltonian) for m in circuit.measurements
            )
```

This and a test to go along with it.

File: tests/devices/test_default_qutrit_mixed.py
Comment (mudit2812) [reply tests/devices/test_default_qutrit_mixed.py:297]: We don't generally inherit from test classes, so I don't personally mind having this duplication in tests. However, in the future, we can consider collecting the tests into one file and parametrizing over the device. This type of duplication is rampant in the `tests/interfaces` module, where the test files for each interface have the same tests, with the only significant difference being the function used to compute gradients/jacobians, as those are dependent on the interface being used.

File: tests/devices/test_default_qutrit_mixed.py
Comment (mudit2812) [inline tests/devices/test_default_qutrit_mixed.py:753]: 🎉

File: tests/devices/test_default_qutrit_mixed.py
Comment (mudit2812) [inline tests/devices/test_default_qutrit_mixed.py:1007]: @albi3ro we want to avoid changing the global seed in tests now right?

File: tests/devices/test_default_qutrit_mixed.py
Comment (mudit2812) [inline tests/devices/test_default_qutrit_mixed.py:1058]: We recently updated how `PRNGKey`s are used for execution. These tests might have to be updated accordingly.

File: pennylane/devices/default_qutrit_mixed.py
Comment (trbromley) [reply pennylane/devices/default_qutrit_mixed.py:77]: Right, it's a question of whether we are tracking for what is actually happening vs what we would want to happen on an optimized hardware device.

We have a similar question in `default.qubit` and solved it in https://github.com/PennyLaneAI/pennylane/pull/4628, with `executions` being the number of hardware executions and `simulations` being the number of times you would simulate.

So in this case, I guess we are correctly reporting `simulations` but `executions` could be lower?

Overall though, I think we should not block this PR - we can create an issue to follow up later.

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:297]: Alright I will leave it to for the future

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:1058]: Would that affect how it is written for this device, do I need to update that?

File: pennylane/devices/default_qutrit_mixed.py
Comment (mudit2812) [reply pennylane/devices/default_qutrit_mixed.py:201]: You're right, nevermind!

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:1010]: I had one idea i can get the state of the rng change the rng then revert to the save state. That or I can set it to none at the end of the test.

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:1064]: Oops I didnt see above, my bad I will look into it

File: doc/releases/changelog-dev.md
Comment (Gabriel-Bottrill) [reply doc/releases/changelog-dev.md]: It seems like that was added in a different PR, I moved the message for this PR.

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:1058]: I updated it to be like how `default_qubit` tests are now.

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:1007]: I undid that because it shows up in a few tests that change global seed.
These tests are similar to the qubit ones and I checked that they are still the same for the qubit version, do I need to change anything?

**Comment by Gabriel-Bottrill** [comment]
Hey @albi3ro, @glassnotes,  @mudit2812, @trbromley. I have made all the requested changes and closed all the conversations, but one. Sorry that my response has been delayed on this one I had to balance a few things.

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:1058]: The updates were seperate from how QutritMixed sampling happens I tried to make the tests more similar but I have to leave some tests out as it will require changing some of the sampling code which I think should be a separate PR. 

Namely I left out `test_prng_key_multi_tapes`  as it does not use a new PRNG and the `test_get_prng_keys` test as that method has not been added

Comment (mudit2812) [reply tests/devices/test_default_qutrit_mixed.py:1058]: Ok, we can cover this in the future in that case.

Comment (mudit2812) [reply tests/devices/test_default_qutrit_mixed.py:1058]: Sounds good to me. In that case, we don't need to add the `_prng_seed` argument, as that is used for resetting the PRNGKey before new executions.

File: pennylane/devices/default_qutrit_mixed.py
Comment (mudit2812) [inline pennylane/devices/default_qutrit_mixed.py:180]: As I said in my other comment, since we're making the PRNGKey changes in the future, we can get rid of these for now.

```suggestion
            self._prng_key = seed
            self._rng = np.random.default_rng(None)
        else:
```

**Review by mudit2812 — APPROVED** [review]
Nice work Gabe! Congrats on getting this in. Just a note: the PRNGKey updates are needed for better support for jax-jit, but we can make it a future improevment.

File: tests/devices/test_default_qutrit_mixed.py
Comment (mudit2812) [reply tests/devices/test_default_qutrit_mixed.py:1007]: Yeah I just checked the `default.qubit` tests and they do the same thing. No changes needed in that acse. Resolving.

File: tests/devices/test_default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply tests/devices/test_default_qutrit_mixed.py:1058]: Good poing i will remove

**Comment by Gabriel-Bottrill** [comment]
@glassnotes, I added tests for integration, I ended up catching a small bug (for if you want to measure more wires than are in the circuit and fixed that.

File: pennylane/devices/default_qutrit_mixed.py
Comment (mudit2812) [inline pennylane/devices/default_qutrit_mixed.py:53]: Why is this here if the `observable_stopping_condition` uses `DefaultQutrit.observables`?

File: pennylane/devices/qutrit_mixed/measure.py
Comment (mudit2812) [inline pennylane/devices/qutrit_mixed/measure.py:156]: Curious why this change was added.

**Review by mudit2812 — COMMENTED** [review]
Tests look good! Just a couple of questions.

File: pennylane/devices/default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply pennylane/devices/default_qutrit_mixed.py:53]: My intention was to support them as I realized they were not supported, I removed the call to `qml.DefaultQutrit.observables` too as this device doesn't currently support Identity

File: pennylane/devices/qutrit_mixed/measure.py
Comment (Gabriel-Bottrill) [reply pennylane/devices/qutrit_mixed/measure.py:156]: It only occurs if the number of wires that have operators is less than number of measured wires and there is a gap in the operator wires ie wire 0 and 2 have operators, but 1 doesn't

File: pennylane/devices/default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply pennylane/devices/default_qutrit_mixed.py:53]: I am just working on this now

File: pennylane/devices/default_qutrit_mixed.py
Comment (mudit2812) [inline pennylane/devices/default_qutrit_mixed.py:66]:
```suggestion
        return all(observable_stopping_condition(observable) for observable in obs.obs)
```

Comment (Gabriel-Bottrill) [reply pennylane/devices/default_qutrit_mixed.py]: Fixed

File: pennylane/devices/default_qutrit_mixed.py
Comment (Gabriel-Bottrill) [reply pennylane/devices/default_qutrit_mixed.py:53]: Added a recursive function to ensure proper observable input

File: doc/releases/changelog-dev.md
Comment (trbromley) [inline doc/releases/changelog-dev.md:429]:
```suggestion
```

File: pennylane/devices/default_qutrit_mixed.py
Comment (trbromley) [inline pennylane/devices/default_qutrit_mixed.py:210]:
```suggestion
            bool: Whether or not a derivative can be calculated provided the given information.
```

File: pennylane/devices/default_qutrit_mixed.py
Comment (trbromley) [inline pennylane/devices/default_qutrit_mixed.py:257]:
```suggestion
            TransformProgram, ExecutionConfig: A transform program that when called returns
            ``QuantumTape`` objects that the device can natively execute, as well as a postprocessing
            function to be called after execution, and a configuration with unset
            specifications filled in.
```
No indent needed, otherwise it shows incorrectly: https://xanaduai-pennylane--5495.com.readthedocs.build/en/5495/code/api/pennylane.devices.default_qutrit_mixed.DefaultQutritMixed.html#pennylane.devices.default_qutrit_mixed.DefaultQutritMixed.preprocess

**Review by trbromley — APPROVED** [review]
Thanks @Gabriel-Bottrill @glassnotes, it's great to have this device added :rocket:

**Review by glassnotes — APPROVED** [review]
Thanks for implementing the final changes @Gabriel-Bottrill !

I think there are a few follow-up things we need to discuss (e.g., more integration tests for the other ML interfaces, and I think there was something else about the PRNG keys?) but the core functionality is there and it should not block completion of the device.

Thanks to everyone on the PL team for helping get this across the finish line!

**Pull request merged**
